### PR TITLE
8244535: JavaDoc search is overly strict with letter case

### PR DESCRIPTION
**Pull request by hns**
This PR softens the case-sensitivity rules in Javadoc searches by including results of case-insensitive search if a case-sensitive yields no or very few results. 

Changes also include some restructuring of the search.js code along with minor changes that improve compliance with the specification (Segments separated by `_` are ranked the same way as segments of CamelCase identifiers, and matches on such segments are ranked lower than matches for whole identifiers, which wasn't always the case previously)

The `TestSearchScript.java` test, which has been disabled with the removal of Nashorn, has been updated to run without failure. In the process, I noticed that the recent IndexBuilder|Writer|Item refactoring, comparators have been unified to the point that the same item order is used in HTML index pages and the search index JSON files. However, there used to be slight differences in the sorting between the two. This change restores the original order by adding two simple search index comparators to `IndexBuilder.java`.

API documentation generated with this PR applied can be viewed and tested here: 
http://cr.openjdk.java.net/~hannesw/8244535/api.01/

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8244535](https://bugs.openjdk.java.net/browse/JDK-8244535): JavaDoc search is overly strict with letter case

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1354/head:pull/1354`
`$ git checkout pull/1354`

### CODE REVIEW

**Comment by pavelrappo** [comment]
@hns, could you recommend (at least several different) example searches that would clearly show the advantages of this proposal over [JDK 15](https://docs.oracle.com/en/java/javase/15/docs/api/index.html)? Although the are examples in tests, they are contrived (naturally). It would be nice to see how this improves search on a real codebase, such as the JDK codebase.

**Review by jonathan-gibbons — APPROVED** [review]
There is a certain amount of "two steps forward, one step back" here.

Yes, this is a good fix to the regression, so approved, and maybe it was an embarrassing convenience that the search test was disabled for the IndexBuilder work, but it is retrograde in that we're going back to a world with different sort orders for the static index pages, and interactive search.

FWIW, I like the conciseness of the change, which also helps highlight when comparators would be different ... i.e. for type elements with the same simple name.  With that in mind, I came up with the "test case" for when the comparators are different: the `ToolProvider` class, which is in both `java.util.spi` and `java.tools`, and these two instances sort differently in the two presentations.

Bottom line: this is a good fix, and a step in the right direction, but we should follow up with another fix to have the static index pages sort the same as the index pages.  I'm guessing that may mean reverting part of this change and improving `mainComparator`.

**Comment by jonathan-gibbons** [comment]
I thought I saw a different order in the type declarations. I'll check 
again.

-- Jon

**Comment by jonathan-gibbons** [comment]
With the change, the behavior with the proposed patch is back to being the same as JDK 15, which is the indirect result of the direct goal to fix a broken test.

However, with the change, the sort order of the two instances of ToolProvider is different in the interactive search and static index.   This is a change from recent behavior in JDK 16 where the sort order is the same. Until this patch is pushed and a new promotion published, you can see the recent JDK 16 behavior here:  https://download.java.net/java/early_access/jdk16/docs/api/index.html

**Comment by hns** [comment]
> @hns, could you recommend (at least several different) example searches that would clearly show the advantages of this proposal over [JDK 15](https://docs.oracle.com/en/java/javase/15/docs/api/index.html)? Although the are examples in tests, they are contrived (naturally). It would be nice to see how this improves search on a real codebase, such as the JDK codebase.

For case insensitivity, you can search for "HTTPC" to find `HttpClient` (or any other name where you get capitalization wrong). 

For the tweaks and fixes in result ranking, you can search for "get". Previously you got several variables with longer names containing "GET" ranked at the top of the result, now it's GET and get() at the top.

I have a few searches I routinely do to detect and understand changes in search behaviour, but honestly I think there is nothing special about these searches. :)
